### DISCUSSION TITLE
Serialize to XML with multiple namespaces

### DISCUSSION THREAD

**Discussion by lppedd** [pdvrieze/xmlutil#48] [Q&A]
Hi! Thanks for this awesome MP library 😄  
How would you write a class that needs to be serialized to

```
<Envelope 
   xmlns="http://schemas.xmlsoap.org/soap/envelope/"
   xmlns:wsse="http://schemas.xmlsoap.org/ws/2003/06/secext"
   xmlns:saml="urn:oasis:names:tc:SAML:1.0:assertion"
   xmlns:wsp="http://schemas.xmlsoap.org/ws/2002/12/policy"
   xmlns:wsu="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd"
   xmlns:wsa="http://schemas.xmlsoap.org/ws/2004/03/addressing"
   xmlns:wssc="http://schemas.xmlsoap.org/ws/2004/04/sc"
   xmlns:wst="http://schemas.xmlsoap.org/ws/2004/04/trust"
>
  ...
</Envelope>
```
All those namespaces are mandatory unfortunately. Even though some of them are unused.

**Comment by pdvrieze** [comment]
Yet another example of a silly XML consumer processing "almostXml". But I assume that your question is not how to use multiple namespaces itself (which is supported as long as you have the correct annotations). Instead I assume you want to inject these namespace attributes. To do so you will have to have a custom serializer (for `Envelope`). The serializer can (more or less) be in line with the autogenerated version, but inside the `encodeStructure` block you would test that the encoder is actually an `XML.XmlOutput` instance. In that case you can get access to the underlying `XmlWriter`. Then you can use `XmlWriter.namespaceAttr` to write your namespace attributes. (you do this before writing any of the contents of the envelope). Note that the serialization should automatically pick up the declared prefixes, rather than using the defined ones.
